[MRG] Add KMM method + example

antoinedemathelin opened the pull request:
Hi everyone,
I open this PR to propose an implementation of the KMM method from the paper https://proceedings.neurips.cc/paper_files/paper/2006/file/a2186aa7c086b46ad4e8bf81e2a3a19b-Paper.pdf as suggested in issue #82.

KMM has the particularity to directly learn the source weights corresponding to the source data Xs given in `fit`. In the `adapt` method, I follow the same idea as for OT. The source data matrix are stored in an attribute `X_source_`, then, in the adapt method, we compare the given X[src_idx] with X_source_ (cf. [OT1](https://github.com/PythonOT/POT/blob/c84ef332ea50a5f94a72f435a7828989c15fa4b5/ot/da.py#L593)), if both are equal, we return the weights learned in fit. If not, we return the weight of the closest neighbor in X_source_, as done for OT mapping, cf. [OT2](https://github.com/PythonOT/POT/blob/c84ef332ea50a5f94a72f435a7828989c15fa4b5/ot/da.py#L612) (note that we use the kernel function to find the closest neighbor).

YanisLalou commented on examples/methods/plot_kmm_reweighting.py:
```diff
@@ -0,0 +1,60 @@
+
+# Author: Antoine de Mathelin
+#
+# License: MIT License
```
We've changed the license a while back I think, now its the BSD 3-Clause

YanisLalou commented on examples/methods/plot_kmm_reweighting.py:
```diff
@@ -0,0 +1,60 @@
+# %%
+# Generate sample bias dataset
+# ----------------------------
+np.random.seed(123)
+Xs = np.random.randn(100, 2) * .5
+Xt = np.random.random((20, 2))
+
+X = np.concatenate((Xs, Xt))
+sample_domain = np.ones(X.shape[0])
+sample_domain[Xs.shape[0]:] *= -2
```
Usually we use the `make_shifted_datasets()` method  to generate our example datasets. Here you can use it with the arg `shift="covariate_shift"`

rflamary commented on examples/methods/plot_kmm_reweighting.py:
```diff
@@ -0,0 +1,65 @@
+Xt, yt = ds.get_domain("t")
+
+# %%
+# Illustration of Importance Weighting
```
Very nice, maybe we could also illustrate the trained predictor with and woithout teh revwieghting? after tgis visualisation?

rflamary commented on skada/_reweight.py:
```diff
@@ -8,8 +8,9 @@
 
 import numpy as np
 from scipy.stats import multivariate_normal
+from scipy.optimize import minimize, LinearConstraint
```
We will have  a conflict between antoines here (@antoinedemathelin VS @antoinecollas ) ;)

rflamary commented on skada/_reweight.py:
```diff
@@ -669,3 +670,272 @@ def KLIEP(
+        Tolerance for the stopping criterion in the optimization.
+    max_iter : int, default=100
+        Number of maximum iteration before stopping the optimization.
+    random_state : int, RandomState instance or None, default=None
```
why random state hare?

rflamary commented on skada/_reweight.py:
```diff
@@ -669,3 +670,272 @@ def KLIEP(
+        )
+        X_source, X_target = source_target_split(X, sample_domain=sample_domain)
+
+        self.weights_ = self._weights_optimization(X_source, X_target)
```
```suggestion
        self.source_weights_ = self._weights_optimization(X_source, X_target)
```
maybe?

rflamary replied:
no problem if you want to keep weights but i think it is more clear

rflamary commented on skada/_reweight.py:
```diff
@@ -669,3 +670,272 @@ def KLIEP(
+                               gamma=self.gamma,
+                               degree=self.degree,
+                               coef0=self.coef0)
+        Ns = Kss.shape[0]
```
Same comment as is did to @antoinecollas , we need a function for QP solver, we will use it many times probably. You should proably work together on this one with one merge before the other

rflamary replied:
Also I'm sorry but you estimate a kernel machine reweighting sum_i w_i k(x,,x_i) my intuition is you should compute this rewieghting instead of juste the weights estimated on the points and then nearest neighbor (let us talk about this in slack)

rflamary commented on skada/_reweight.py:
```diff
@@ -669,3 +670,272 @@ def KLIEP(
+                source_weights = self.weights_
+            else:
+                # get the nearest neighbor in the source domain
+                K = pairwise_kernels(X[source_idx], self.X_source_,
```
Same cmment again, why compute a kerel when distane is enough?

rflamary reviewed (commented):
Left a few comments. We need to disus with @antoinecollas about a few things to avoid conflicts and repeated code

antoinedemathelin commented:
Hi everyone,
I added a `qp_solve` function to factorize the qp solver. I need to create specific test for it, I think.
@rflamary I added the `smooth_weights` parameter as we discussed.

rflamary reviewed (commented):
This is nice. We are missing a few tests on the new solver function and I would illustrate the final classifier and weights for smooth_weights and not smooth weights in the example (r elese nobody will use it)

rflamary commented on examples/methods/plot_kmm_reweighting.py:
```diff
@@ -0,0 +1,107 @@
+"""
+Kernel Mean Matching
+====================
+This example illustrates the use of KMM method [1] to correct covariate-shift.
```
Small rendering problem in the example

![image](https://github.com/scikit-adaptation/skada/assets/1386873/579c7edc-e334-47d9-8bff-2a70ad6da784)

rflamary commented on skada/_reweight.py:
```diff
@@ -669,3 +669,266 @@ def KLIEP(
+        self.tol = tol
+        self.max_iter = max_iter
+        self.smooth_weights = smooth_weights
+        self.random_state = random_state
```
random state is set but never used... remove it?

rflamary commented on skada/_reweight.py:
```diff
@@ -669,3 +669,266 @@ def KLIEP(
+        if self.eps is None:
+            eps = (np.sqrt(Ns) - 1) / np.sqrt(Ns)
+        else:
+            eps = self.eps
```
line ,never tested

rflamary commented on skada/_reweight.py:
```diff
@@ -669,3 +669,266 @@ def KLIEP(
+               and not self.smooth_weights):
+                source_weights = self.weights_
+            else:
+                K = pairwise_kernels(X[source_idx], self.X_source_,
```
not covered by tests

rflamary commented on skada/_reweight.py:
```diff
@@ -669,3 +669,266 @@ def KLIEP(
+        self.random_state = random_state
+
+        if kernel not in KERNEL_PARAMS:
+            raise ValueError("`kernel` argument should be included in %s,"
```
Value Error not tested

rflamary reviewed (commented):
I'm sorry @antoinedemathelin but O found a few stuff remaining see below. This is nearly perfect.
